File: framework/elsa/fit-elsa-react/src/components/base/validator.js
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) 2025 Huawei Technologies Co., Ltd. All rights reserved.
 *  This file is a part of the ModelEngine Project.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

class Validator {
  constructor(node) {
    this.node = node;
  }

  validate() {
    throw new Error('Method \'validate()\' must be implemented.');
  }
}

export class FormValidator extends Validator {
  validate() {
    return new Promise((resolve, reject) => {
      try {
        this.node.validateForm().then(resolve).catch(reject);
      } catch (error) {
        reject({
          errorFields: [{
            errors: [error.message],
            name: 'node-error',
          }],
        });
      }
    });
  }
}

export class NormalNodeConnectorValidator extends Validator {
  validate() {
    const nextEvents = this.node.getNextRunnableEvents();
    const i18n = this.node.graph.i18n;
    if (nextEvents.length !== 1) {
      return Promise.reject({
        errorFields: [{
          errors: [`${i18n?.t('node') ?? 'node'} ${this.node.text} ${i18n?.t('problemWithConnection') ?? 'problemWithConnection'}`],
          name: 'link-error',
        }],
      });
    }
    return Promise.resolve();
  }
}

export class ConditionNodeConnectorValidator extends Validator {
  validate() {
    const nextEvents = this.node.getNextRunnableEvents();
    const i18n = this.node.graph.i18n;
    if (nextEvents.length !== this.node.getBranches().filter(b => b.runnable).length) {
      return Promise.reject({
        errorFields: [{
          errors: [`${i18n?.t('node') ?? 'node'} ${this.node.text} ${i18n?.t('problemWithConnection') ?? 'problemWithConnection'}`],
          name: 'link-error',
        }],
      });
    }
    return Promise.resolve();
  }
}

export class EndNodeConnectorValidator extends Validator {
  validate() {
    const nextEvents = this.node.getNextRunnableEvents();
    const i18n = this.node.graph.i18n;
    if (nextEvents.length !== 0) {
      return Promise.reject({
        errorFields: [{
          errors: [`${i18n?.t('node') ?? 'node'} ${this.node.text} ${i18n?.t('problemWithConnection') ?? 'problemWithConnection'}`],
          name: 'link-error',
        }],
      });
    }
    return Promise.resolve();
  }
}

export class KnowledgeRetrievalValidator extends Validator {
  validate() {
    const jadeConfig = this.node.drawer.getLatestJadeConfig();
    const i18n = this.node.graph.i18n;

    // 校验搜索参数.
    const option = jadeConfig.inputParams.find(ip => ip.name === 'option');
    if (option.value.length === 0) {
      return Promise.reject({
        errorFields: [{
          errors: [`${this.node.text} ${i18n?.t('noSearchOption') ?? 'noSearchOption'}`],
          name: 'search-args-error',
        }],
      });
    }

    return Promise.resolve();
  }
}

File: app-builder/waterflow/java/waterflow-service/src/main/java/modelengine/fit/waterflow/flowsengine/domain/flows/validators/Validator.java
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) 2025 Huawei Technologies Co., Ltd. All rights reserved.
 *  This file is a part of the ModelEngine Project.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

package modelengine.fit.waterflow.flowsengine.domain.flows.validators;

import modelengine.fit.waterflow.exceptions.WaterflowParamException;
import modelengine.fit.waterflow.flowsengine.domain.flows.definitions.FlowDefinition;

/**
 * 流程定义校验接口
 *
 * @author 高诗意
 * @since 2023/08/29
 */
public interface Validator {
    /**
     * 校验流程定义的合法性
     * 当校验不通过时，抛出运行时异常{@link WaterflowParamException}
     *
     * @param flowDefinition 流程定义实体
     */
    void validate(FlowDefinition flowDefinition);

    /**
     * 校验分页参数的合法性
     *
     * @param offset 分页参数：偏移
     * @param limit 分页参数：条数
     */
    void validatePagination(int offset, int limit);
}
